[Docs] Adds mustache dependency search templates (#95118)
elastic/elasticsearch: docs/reference/search/search-your-data/search-template.asciidoc

To create or update a search template, use the <<create-stored-script-api,create
stored script API>>.

The request's `source` supports the same parameters as the
<<search-search-api-request-body,search API>>'s request body. `source` also

> supports https://mustache.github.io/[Mustache] variables and has https://github.com/elastic/elasticsearch/blob/main/modules/lang-mustache/build.gradle/[dependency] on https://github.com/spullara/mustache.java[mustache.java]. Typically https://mustache.github.io/[Mustache] variables are enclosed in

Reviewer: ```suggestion
supports https://mustache.github.io/[Mustache] variables and has a https://github.com/elastic/elasticsearch/blob/main/modules/lang-mustache/build.gradle/[dependency] on the open source project https://github.com/spullara/mustache.java[mustache.java]. 

Typically https://mustache.github.io/[Mustache] variables are enclosed in
```

double curly brackets: `{{my-var}}`. When you run a templated search, {es}

> replaces these variables with values from `params`. To learn more about mustache syntax - see http://mustache.github.io/mustache.5.html/[Mustache.js manual]

Reviewer: ```suggestion
replaces these variables with values from `params`. To learn more about mustache syntax - see http://mustache.github.io/mustache.5.html[Mustache.js manual]
```

To create or update a search template, use the <<create-stored-script-api,create
stored script API>>.

The request's `source` supports the same parameters as the
<<search-search-api-request-body,search API>>'s request body. `source` also

> supports https://mustache.github.io/[Mustache] variables and has a https://github.com/elastic/elasticsearch/blob/main/modules/lang-mustache/build.gradle/[dependency] on the open source project https://github.com/spullara/mustache.java[mustache.java]. 

Reviewer 2: Why is it relevant for our documentation that we have a gradle dependency on Mustache? Can we either omit that fact or elaborate on it?

Author: I think the main intention was to let user know the dependency. And assume that user could read more about Mustache in their GitHub repo

> Typically https://mustache.github.io/[Mustache] variables are enclosed in

Reviewer 2: Was the spirit of this issue to elaborate a bit more on some of the commonly used mustache features?

We already mention that variables are enclosed in double curly brackets. While we don't want to copy all of their documentation, would it make sense to mention that other functionality in Mustache is supported, such as iterating over arrays, conditionals, inverted conditionals, and delimeters? 

Another thing that could be worth noting (maybe?) _for 8.8_ is that there is a known issue with trailing commas in Mustache lists because it creates invalid parsing. This requires a workaround in order to use effectively. 

Perhaps a visual example of a mustache template that used variables and conditionals here would be more useful than explaining it in text. 

I like that you link to the full documentation if they'd like to know more.

Reviewer 3: > Perhaps a visual example of a mustache template that used variables and conditionals here would be more useful than explaining it in text.

I agree an example would be great. _Show_ (rather than tell) how an Elasticsearch user may want to use mustache. Let the mustache docs provide all the details.

To create or update a search template, use the <<create-stored-script-api,create
stored script API>>.

The request's `source` supports the same parameters as the
<<search-search-api-request-body,search API>>'s request body. `source` also

> supports https://mustache.github.io/[Mustache] variables and has a https://github.com/elastic/elasticsearch/blob/main/modules/lang-mustache/build.gradle/[dependency] on the open source project https://github.com/spullara/mustache.java[mustache.java].

Reviewer 4: Mentioning the gradle dependency assumes that users are familiar with Gradle, which is not necessarily the case, even for developers. I could see this being useful if the gradle dependency needed to be manually provided, but given that it's pre-installed and part of the default distribution, I find this mention redundant and potentially confusing. Fine specifying that we rely on an open source project and linking to it, but I would remove the gradle mention.

  }
}
----
[[search-template-with-mustache-examples]]
=== Search template examples with Mustache

The mustache templating language defines various tag types you can use within templates. The following sections describe some of these tag types and provide examples of using them in {es} <<search-template, search templates>>.

[discrete]
[[search-template-mustache-variable]]
=== Mustache variables

> Mustache tags are typically enclosed in double curly brackets, A mustache variable : `{{my-variable}}` is a type of mustache tag which checks for a matching `my-variable` key in your search template `params` and replaces template with the value set for `my-variable`

Reviewer 5: ```suggestion
Mustache tags are typically enclosed in double curly brackets, A mustache variable : `{{my-variable}}` is a type of mustache tag which checks for a matching `my-variable` key in your query `params` and replaces template with the value set for `my-variable`
```

Reviewer 5: Maybe it's better to say "query params" here?

Author: I think just `params` would be fine? Similar to how current search template documentation refers to `params` in [create search template](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-template.html#create-search-template)

So something like :
``` 
... checks for a matching `my-variable` key in your `params`...
```

My reasoning: `params` is optional tag set outside `query` but within `source` in search template.  WDYT?

      "query": {
        "match": {
          "message": "{{query_string}}"
        }
      },
      "from": "{{from}}",
      "size": "{{size}}"
    },
    "params": {
      "query_string": "My query string"
    }
  }
}

----

Testing `my-search-template` with different `params`

[source,console]

----

POST _render/template
{
  "id": "my-search-template",
  "params": {
    "query_string": "hello world",
    "from": 20,
    "size": 10
  }
}

----

When rendered, the template outputs as:
[source,console-result]

----

{
  "template_output": {
    "query": {
      "match": {
        "message": "hello world"
      }
    },
    "from": "20",
    "size": "10"
  }
}

----

If your search template doesn't pass a value to your `query_string` the message would be replaced with a empty string.

[source,console]

----

POST _render/template
{
  "id": "my-search-template",
  "params": {
    "from": 20,
    "size": 10
  }
}

----

When rendered, template outputs as:
[source,console-result]

----

{
  "template_output": {
    "query": {
      "match": {
        "message": ""
      }
    },
    "from": "20",
    "size": "10"
  }
}

----

[discrete]
[[search-template-sections]]
==== Sections

> Sections are also a type of Mustache tags and are useful when you have to loop through nested keys. A sections begins with `{{#my-section-variable}}` and ends with `{{\my-section-variable}}`

Author: Good catch. Sections are similar to conditionals and we could loop through the nested keys. But, the example provided for `sections` doesn't show looping. But I do show the example in LIST. 

Anyway, I guess I will rephrase the sentence a bit and  may be alter example to avoid confusions.

[source,console]

----

POST /_render/template?pretty
{
  "source": {
    "query": {
      "match": {
        "message": "{{^name_exists}}Hello World{{/name_exists}}"
      }
    }
  },
  "params": {
     "name_exists": false
  }
}

----

They can also be combined with  <<search-template-use-conditions,conditions>> and <<search-template-set-default-values,default values>>.

For example, in the following search template, if `name_exists` is true, value of `{{query_string}}` is replaced else is set to default value `World`.

[source,console]

----

POST /_render/template?pretty
{
  "source": {
    "query": {
      "match": {
        "message": "Hello {{#name_exists}}{{query_string}}{{/name_exists}}{{^name_exists}}World{{/name_exists}}"
      }
    }
  },
  "params": {
    "query_string": "Kimchy",
     "name_exists": true
  }
}

----

When rendered, template output:

[source,console-result]

----

{
  "template_output": {
    "query": {
      "match": {
        "message": "Hello Kimchy"
      }
    }
  }
}

----

[discrete]
[[search-template-lists]]
==== Lists

> You can pass list of objects and loop over each item in your search template

Reviewer 5: ```suggestion
You can pass a list of objects and loop over each item in your search template.
```

For example, following search template combines <<search-template-sections,sections>> and matches all the usernames.

[source,console]

----

PUT /_scripts/my-search-template?pretty
{
  "script": {
    "lang": "mustache",
    "source": {
      "query":{
        "multi-match":{
          "query": "{{query_string}}",
          "fields": """[{{#text_fields}}{{user_name}},{{/text_fields}}]"""
        }
      },
      "params":{
        "query_string": "My query string",
        "text_fields": [ { "user_name": "John" }, {"user_name": "kimchy"}]
      }
    }
  }
}

----

When rendered, template outputs:

> [source,console]
> 
> 
> ----
> {
>   "template_output": {
>     "query": {
>       "multi-match": {
>         "query": "My string",
>         "fields": "[John,kimchy,]"
>       }
>     },
>     "params": {
>       "query_string": "My query string",
>       "text_fields": [
>         {
>           "user_name": "John"
>         },
>         {
>           "user_name": "kimchy"
>         }
>       ]
>     }
>   }
> }
> 
> ----

Reviewer 5: ```suggestion
[source,console-result]
----
{
  "template_output": {
    "query": {
      "multi-match": {
        "query": "My string",
        "fields": "[John,kimchy,]"
      }
    },
    "params": {
      "query_string": "My query string",
      "text_fields": [
        {
          "user_name": "John"
        },
        {
          "user_name": "kimchy"
        }
      ]
    }
  }
}
----
```

NOTE: The above will cause trailing comma issue, which cause invalid json. A workaround would be including <<search-template-inverted-section,inverted section>> and add a variable to make sure it's last item in array.

For example

[source,console]

----

PUT /_scripts/my-search-template?pretty
{
  "script": {
    "lang": "mustache",
    "source": {
      "query":{
        "multi-match":{
          "query": "{{query_string}}",
          "fields": """[{{#text_fields}}{{user_name}}{{^last}},{{/last}}{{/text_fields}}]"""
        }
      },
      "params":{
        "query_string": "My query string",
        "text_fields": [
          {
            "user_name": "John",
            "last": false
          },
          {
            "user_name": "kimchy",
            "last": true

          }
        ]
      }
    }
  }
}

----

When rendered the template outputs:

> [source,console]
> 
> ----
> 
> 
> {
>   "template_output": {
>     "query": {
>       "multi-match": {
>         "query": "My string",
>         "fields": "[welcome,kimchy]"
>       }
>     },
>     "params": {
>       "query_string": "My query string",
>       "text_fields": [
>         {
>           "user_name": "John",
>           "last": false
>         },
>         {
>           "user_name": "kimchy",
>           "last": true
>         }
>       ]
>     }
>   }
> }
> 
> ----

Reviewer 5: ```suggestion
[source,console-result]
----
{
  "template_output": {
    "query": {
      "multi-match": {
        "query": "My string",
        "fields": "[welcome,kimchy]"
      }
    },
    "params": {
      "query_string": "My query string",
      "text_fields": [
        {
          "user_name": "John",
          "last": false
        },
        {
          "user_name": "kimchy",
          "last": true
        }
      ]
    }
  }
}
----
```

[discrete]
[[search-template-set-delimiter]]
==== Set delimiter

> You can change the default demiliter: double curly brackets `{{my-variable}}` to any custom delimiter in your search template
> 
> For example, following search template changes the default demiliter to single round bracket `(my-variable)`

Reviewer 5: ```suggestion
You can change the default delimiter: double curly brackets `{{my-variable}}` to any custom delimiter in your search template.

For example, the following search template changes the default delimiter to a single round bracket `(my-variable)`.
```

> [source,console]
> 
> 
> ----
> 
> 
> PUT /_scripts/my-search-template?pretty
> {
>   "script": {
>     "lang": "mustache",
>     "source":
>     """
>     {
>       "query": {
>         "match": {
>            {{=( )=}}
>           "message": "(query_string)"
>           (={{ }}=)
>         }
>       }
>     }
>     """,
>     "params": {
>       "query_string": "Hello"
>     }
>   }
> }
> 
> ----

Reviewer 5: ```suggestion
[source,console]
----
PUT /_scripts/my-search-template?pretty
{
  "script": {
    "lang": "mustache",
    "source":
    """
    {
      "query": {
        "match": {
           {{=( )=}}
          "message": "(query_string)"
          (={{ }}=)
        }
      }
    }
    """,
    "params": {
      "query_string": "Hello"
    }
  }
}
----
```

When rendered, template outputs:

> [source,console]
> 
> 
> ----
> 
> {
>   "template_output": {
>     "query": {
>       "match": {
>         "message": "Hello world"
>       }
>     }
>   }
> }
> ----

Reviewer 5: ```suggestion
[source,console-result]
----
{
  "template_output": {
    "query": {
      "match": {
        "message": "Hello world"
      }
    }
  }
}
----

```

[discrete]
[[search-template-lambdas]]
==== Lambdas

> Elasticsearch has pre-built custom functions to support converting the text into a specific format.

Reviewer 5: ```suggestion
{es} has pre-built custom functions to support converting the text into a specific format.
```

> To Learn more about usage of mustache lambdas, checkout examples in <<search-template-url-encode-strings,Url encode strings>> <<search-template-concatenate-values,Concatenate values>> and <<search-template-convert-json, Convert to json>>

Reviewer 5: ```suggestion
To Learn more about usage of mustache lambdas, check out the examples in <<search-template-url-encode-strings,Url encode strings>>, <<search-template-concatenate-values,Concatenate values>>, and <<search-template-convert-json, Convert to json>>.
```

[discrete]
[[search-template-unsupported-features]]
==== Unsupported features

> * Elasticsearch doesn't support Partials.

Reviewer 5: ```suggestion
* {es} doesn't support partials.
```

[source,console]
----
PUT _scripts/my-search-template?pretty
{
  "script": {
    "lang": "mustache",
    "source":
    """
    {
      "query": {
        "match": {
           {{=( )=}}
          "message": "(query_string)"
          (={{ }}=)
        }
      }
    }
    """,
    "params": {
      "query_string": "Hello"
    }
  }
}
----
Testing the template with new delimiter:

[source,console]
----
POST _render/template
{
  "id": "my-search-template",
  "params": {
    "query_string": "hello world"
  }
}
----

When rendered, template outputs:

[source,console-result]
----
{
  "template_output": {
    "query": {
      "match": {
        "message": "Hello world"
      }
    }
  }
}
----
[discrete]
[[search-template-lambdas]]
==== Lambdas
{es} has pre-built custom functions to support converting the text into a specific format.

To Learn more about usage of mustache lambdas, check out the examples in <<search-template-url-encode-strings,Url encode strings>>, <<search-template-concatenate-values,Concatenate values>>, and <<search-template-convert-json, Convert to json>>.

[discrete]
[[search-template-unsupported-features]]

> ==== Unsupported features

Reviewer 3: Suggested revision for this section:

```asciidoc
==== Unsupported mustache features

The following mustache features are not supported in {es} search templates:

* Partials
```
